### PR TITLE
Some minor search result style changes + datelastx

### PR DESCRIPTION
**Pull request by diabolusss**
Hello!
Maybe you'll find some changes interesting to add to your repo.
 - datelastx - instead of using constants, i think it's better to allow user to control value he wants to enter, i.e. datelastx:7 will search for last 7 hours;
 - on first open of search page it loads full log, i've limited it to the last 6 hours;
 - less relevant changes - i've changed full row style to match severity value;
 - i've added TRACE level which is useful for logs that are more verbose than DEBUG. Of course, it's not listed in syslog rfc, yet rsyslog allows modifying templates and some apps have this level.

### CODE REVIEW

File: src/include/constants_general.php
Comment (alorbach) [inline src/include/constants_general.php:172]: This priority is actually impossible to map because priority is a 3 Bit and facility a 5 bit Number both combined in a one 8 Bit number. So I would like the SYSLOG_TRACE modification bneing removed from this PR.

File: src/include/functions_common.php
Comment (alorbach) [inline src/include/functions_common.php:514]: I would rather recommend to add a configuration variable for this and write the default of the configuration variable into $_SESSION['AUTORELOAD_ID']. Autoreload might be desired in a later step of loganalyzer setup but not from the beginning.

https://github.com/rsyslog/loganalyzer/blob/master/src/include/config.sample.php

File: src/index.php
Comment (alorbach) [inline src/index.php:132]: Same as with AUTORELOAD_ID, have a configuration variable defined in the config.sample.php template, and write this default into $_SESSION['SESSION_MAXIMIZED']. Also you need to define a default for SESSION_MAXIMIZED if the configuration variables is not defined.

**Review by alorbach — COMMENTED** [review]
Thanks for this huge PR, there are indeed useful enhancments. However I have a few requests for changes before we can merge this PR, please take a look.

File: src/include/constants_general.php
Comment (diabolusss) [reply src/include/constants_general.php:172]: Ok, i'll try to figure it out how to remove it without affecting my own branch. Maybe i'll need to create a separate branch...
Most of the suggestions i've fixed, yet right now this one is a dealbreaker.

Comment (rgerhards) [reply src/include/constants_general.php:172]: Just out of curiosity: why do you need non-existing SYSLOG_TRACE? Any non-syslog sources inside the database?

Comment (diabolusss) [reply src/include/constants_general.php:172]: Well, I route logs from my router (Mikrotik Chateau) to rsyslog and view them using loganalizer.
RouterOS have a lot of  [topics](https://wiki.mikrotik.com/wiki/Manual:System/Log) instead of standard log levels, which is cool, but needs special handling. Of course, it can format logs in bsd compliant format (as defined in RFC 3164), but it's not informative enough. 
So, there are topics like "packet, raw" and maybe some other very verbose, that floods logs very fast if enabled. And there is "debug" topic (and maybe some other) that is less verbose. Obviously, I want to separate them out, and here comes in TRACE level. Please, see [Severity Occurences chart](https://imgur.com/vJe6CNy.png).

To mark needed log messages as TRACE i use rsyslog useful filters. 

Moreover, i believe, loganalizer should support such levels, at least, because it supports apache logs which in turn have [trace log level](https://drill.apache.org/docs/logging-and-tracing/). But, IMHO, supported levels should be configurable for each source separately (because rsyslog allows to override real values). 

By the way, these tools are awesome, a huge thanks to you and your team for developing them and making open source!

Comment (alorbach) [reply src/include/constants_general.php:172]: Ok I understand what you mean. Technically I only have a problem mapping the TRACE level as Syslog Priority which is not possible from a protocol point of view. But of course I understand that you can put more values into a database syslog source for the syslog severity.

If you could renamed SYSLOG_TRACE into "EXTENDED_TRACE" for example (open for other suggestions) and give it a number above 255, it would be clearly separated from Syslog Priorities (0-7) and all your filtering enhancements would still work.

Comment (diabolusss) [reply src/include/constants_general.php:172]: I suppose "CUSTOM_TRACE"  will be good for this purpose (or maybe "SEVERITY_TRACE"). 
However, do we really need to change its number to above 255? I don't think that syslog priority matrix will be changed in the near future or ever. So, no syslog severity level will ever collide with trace level mapped in loganalyzer. Is LogAnalyzer somewhere recalculating priority or severity values from logs?

Please, don't understand me wrong, I'm not against that, I just don't get how this number affects loganalyzer. If LogAnalyzer is not doing any math (at least i haven't found that in code, please, point me if there is), but using values as they are provided, i see no problem here. As far as i understand, LogAnalyzer is a journal viewer, not recorder or even redactor. (that's why these levels shouldn't be constants, but configurable)
Please, elaborate your point a bit more in this context.

**Comment by davidelang** [comment]
the problem is that the severity and faility get combined, and only 3 bits are 
allocated for severity.

so facilty user (1) + severity Emergency (0) would map the same as facility 
kern (0) + serverity trace (8)

by making your trace be something > 255 it guarantees that there is never going 
to be a collision.

David Lang

File: src/include/constants_general.php
Comment (alorbach) [reply src/include/constants_general.php:172]: The Syslog Severity is stored as 32bit signed integer in the database, so any number from -2147483648 to +2147483648 would be valid. Anyway I would be fine with CUSTOM_TRACE as it clearly separates it from the constant syslog priorities. 

And yes technically it is vey unlikely that those values will every be changed, so I am ok with using 8.

**Comment by diabolusss** [comment]
> the problem is that the severity and faility get combined, and only 3 bits are allocated for severity. so facilty user (1) + severity Emergency (0) would map the same as facility kern (0) + serverity trace (8) by making your trace be something > 255 it guarantees that there is never going to be a collision. David Lang
> […](#)
> On Wed, 11 Jan 2023, Vitaly wrote: However, do we really need to change its number to above 255? I don't think that syslog priority matrix will be changed in the near future or ever. So, no syslog severity level will ever collide with trace level mapped in loganalyzer. **Is LogAnalyzer somewhere recalculating priority or severity values from logs?** Please, don't understand me wrong, I'm not against that, I just don't get _how this number affects loganalyzer._ **If LogAnalyzer is not doing any math (at least i haven't found that in code, please, point me if there is), but using values as they are provided, i see no problem here.** As far as i understand, LogAnalyzer is a journal viewer, not recorder or even redactor. (that's why these levels shouldn't be constants, but configurable) Please, elaborate your point a bit more in this context.

Thanks, that point was already mentioned by @alorbach and I understand it. But it's only true if someone will ever decide to use CUSTOM_TRACE outside LogAnalyzer which eventually should lead to syslog priority matrix change. Additionally, collision to happen LogAnalyzer should somewhere recalculate severity from received priority number. If you mean that, then i understand your worries. (like i've said, i'm not against, i just want to understand why)

Anyway, I'll set it to 256 (or maybe 1792?).

**Comment by rgerhards** [comment]
> Anyway, I'll set it to 256 (or maybe 1792?).

IMHO using 8 is secure. I remember what happend when we the last time tried to fix the too-few severity values (in RFC424) - strong objection. So I tend to say that Severity 0..7 will stick for at least another 20 years, probably longer. Even if 8+ get finally approved, it would not really hurt. Severity values have no strict semantics and so at most you would have potentially others using 8 as well.

Just my 2cts - I am fine with any of the solutions floating around now.

**Comment by rgerhards** [comment]
@diabolusss sry, can not act as reviewer as I do not know the code

File: src/classes/logstream.class.php
Comment (alorbach) [inline src/classes/logstream.class.php:1384]: only cosmetic, could you remove the extra new line at the end? This can cause problems when headers have not been send yet.

File: src/include/config.sample.php
Comment (alorbach) [inline src/include/config.sample.php:134]: I did not think about the changed search URLs yet. This introduces a problem for existing installations where config.php already exists with ´the old URL format. The change will break the URLs for all existing installations

My Solution proposal: keep the original filter "datelastx" as it is but add a new one like "datelastxx" which uses the new filter behavior, and make it default in the sample configuration. It would not break any existing installations.

**Review by alorbach — CHANGES_REQUESTED** [review]
Everything is fine so far (Y) and as I see the enhancements have even increased.
However there is only one single issue left with the DATEMODE_LASTX filter, as in my inline comment this change will cause existing default search urls to break (I didnt think about it last time). So I would propose to keep the existing DATEMODE_LASTX filter and add a new one called DATEMODE_LASTXX for example which implements your new handling and make it default.

File: src/include/config.sample.php
Comment (diabolusss) [reply src/include/config.sample.php]: @alorbach fix is ready

**Review by alorbach — APPROVED** [review]
From what I see it looks goods.
Can this PR be seen as finished @diabolusss ? 
Then I would perform a last test locally to view the changes in browser before merging the PR.

**Comment by diabolusss** [comment]
> From what I see it looks goods. Can this PR be seen as finished @diabolusss ? Then I would perform a last test locally to view the changes in browser before merging the PR.

Of course, however, i don't know the correct way to do that. I suppose a "Close" button will just fail it!? (Sorry, i don't use git a lot)

**Comment by alorbach** [comment]
> 

Don't worry, once I am done confirming my local test installation, I will merge the PR.

**Comment by diabolusss** [comment]
> > 
> 
> Don't worry, once I am done confirming my local test installation, I will merge the PR.

That's great, thank you.

**Comment by diabolusss** [comment]
Hi, @alorbach. I've prepared a new commit. It contains 'suppress duplicate messages' feature fix, and a few other novelties. Should i wait for merge before committing it to allow making separate tests?

**Comment by alorbach** [comment]
lets wait for the merge and then you can create a new PR, otherwise this PR becomes little hard to read ;)

**Comment by alorbach** [comment]
Technically the PR is fine, but there are some css style issues breaking visibility in some cases.
 I can fix those if you want, and merge the PR once the CSS is being fixed. THen you can rebase your next PR once this PR is merged - i have the setup for testing ready.

image

image

**Comment by diabolusss** [comment]
> Technically the PR is fine, but there are some css style issues breaking visibility in some cases. I can fix those if you want, and merge the PR once the CSS is being fixed. THen you can rebase your next PR once this PR is merged - i have the setup for testing ready.

It will be very kind of you, because, while i've a clue where to check those, i don't have a lot of experience in style sheet debugging.

**Comment by alorbach** [comment]
I will have to merge the PR first and apply the css changes / fixes in a separated PR, I am not able to edit the PR from command line only by github.

**Comment by alorbach** [comment]
@diabolusss I have finished the CSS fixes and some other minor issues that came up during testing. 
Please rebase master based on the our repository: 
https://github.com/rsyslog/loganalyzer

For further changes, it would be great if you could use your own custom branches in future (like "git checkout -b myownbranch"), this avoids merging / rebasing when our master branch has been going forward during meantime.

**Comment by diabolusss** [comment]
> @diabolusss I have finished the CSS fixes and some other minor issues that came up during testing. Please rebase master based on the our repository: https://github.com/rsyslog/loganalyzer
> 
> For further changes, it would be great if you could use your own custom branches in future (like "git checkout -b myownbranch"), this avoids merging / rebasing when our master branch has been going forward during meantime.
Thank you, i'll follow your tip.

**Pull request merged**
